### ISSUE TITLE
GDScript: "while true:" block with no "break" statements inside and at least one "return" statement should count as a return statement for the purpose of 

### ISSUE THREAD

**Issue by BenLubar** [godotengine/godot-proposals#10799]
### Describe the project you are working on

Porting a card game to Godot, but the specific type of game is irrelevant for this proposal.

### Describe the problem or limitation you are having in your project

Best way I can explain this is a screenshot of what currently happens with a slightly contrived code example:

![image](https://github.com/user-attachments/assets/d1bebeb2-94ef-4238-afd7-0b1dfacc2ebc)

```gdscript
func count_number_of_lines(file: FileAccess) -> int:
	var lines := 0
	while true:
		if file.eof_reached():
			return lines

		file.get_line()
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This proposal would remove the need for a dummy return statement at the end of the function with a note that it's actually unreachable.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

In the exact situation where:

1. A code path within a function contains a `while` loop with a condition of the boolean literal `true`.
2. The `while` block does not contain any `break` statements. (`break`s in nested loops are ignored for this, but `break`s in other nested blocks like `if` are counted)
3. There is at least one `return` statement within the loop (including in nested loops).

The code validator would treat the `while true:` block as an arbitrary `return` statement from within it for the purpose of determining whether all code paths return a value.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, the point of this proposal is to avoid having to end a function with unreachable code. The workaround (which is the current situation in its entirety) is to add an unreachable return statement of the correct type and optionally annotate it with a comment, like so:

```godot
# unreachable
return null
```

### Is there a reason why this should be core and not an add-on in the asset library?

This is a change to the GDScript code validator in the editor.

**Comment by AThousandShips** [comment]
This isn't guaranteed to ever return, see the "halting problem", and I'd say it's bad practice generally to write functions like this

But it comes down to the level of analysis and how complex we want to make it, and how much it's worth it

Almost all of the time this kind of function is not correct, the so unless we *also* add analysis for potential dangerous loops I don't think we need to treat this case special, and it's a good indicator to the user that the code is something you should pay attention to

Tl;Dr;
Either, IMO, we should have a static analysis feature that earns about potentially unsafe loops as well, or this edge case should be treated no different than other halting problem cases

**Comment by BenLubar** [comment]
Solving the halting problem is not the purpose of the "not all code paths return a value" warning. For example, this function has no warnings despite the fact that code execution will not resume after it runs:

```gdscript
func exit_the_game() -> String:
    get_tree().quit()
    return "unreachable"
```

The purpose of the warning is to make sure that all paths where the function *can* return a value are returning an explicitly specified value and not the `null` that reaching the end of a function without an explicit `return` statement would return.

In this specific case, it is absolutely possible to determine that the function cannot return a value without hitting an explicit return statement, using the rules I specified.

It's quite common to have a loop with an exit condition that needs to be checked between the start and the end of an iteration.

**Comment by AThousandShips** [comment]
I would consider an infinite loop that isn't guaranteed to exit would be considered a code path that doesn't return, but the solution wouldn't be to add a return statement after the loop, but the code in question isn't guaranteed to ever return, and especially static analysis can't check that

What's the difference between your code and:
```gdscript
var do_exit = false

func count_number_of_lines(file: FileAccess) -> int:
	var lines := 0
	while true:
		lines += 1

		if do_exit:
			return lines

		file.get_line()
```

But I can agree that the loop could be treated as a complete path, but I'd say it should be done carefully as loop analysis is a messy question

**Comment by BenLubar** [comment]
This function should also not result in the warning being discussed because there is no path through the function even assuming every branch is reachable where the function reaches the end without an explicit return statement.

```gdscript
func count_number_of_lines(_file: FileAccess) -> int:
	while true:
		pass
```

These equivalent functions in other languages also result in no warnings. Making the condition of the loop a variable does result in a warning or compile error:

```go
// Go 1.23
func foo() int {
	for {}
}
```

```c++
// gcc 14.2 -Wall -Wextra -Werror -pedantic
int foo() {
    while (true);
}
```

```rust
# Rust 1.81.0
fn foo() -> i32 {
    loop {}
}
```

**Comment by Mickeon** [comment]
CC @dalexeev 

A bit neutral on this. If it's an easy fix I welcome it, but if it requires re-engineering the existing parser, it's not worth it in my opinion.

**Comment by dalexeev** [comment]
This is possible to solve, but requires more advanced control flow static analysis. See also:

* godotengine/godot#71763
* godotengine/godot#86521
* godotengine/godot#92341
* #8994

The error "Not all code paths return a value" should only be produced if there is a block with no outgoing edges and no return statement. In this example, the control flow graph looks like this:

![](https://github.com/user-attachments/assets/4d0408bf-cac4-4c84-8e73-ed3fc7f92d2b)

The only block with no outgoing edges (block 3) has a return statement.

The question also arises whether we should ignore unreachable/dead code without return statement.

**Comment by Xtarsia** [comment]
```gdscript
func count_number_of_lines(file: FileAccess) -> int:
	var lines := 0
	while true:
		lines += 1

		if file.eof_reached():
			return lines

		file.get_line()
	return ERROR
```

regardless of coding practice, wouldn't adding this return resolve the problem for OP?
